File: js/src/gdb/tests/test-JSObject.cpp
#include "gdb-tests.h"
#include "jsapi.h"
#include "js/GlobalObject.h"
#include "js/Object.h"  // JS::GetClass

FRAGMENT(JSObject, simple) {
  AutoSuppressHazardsForTest noanalysis;

  JS::Rooted<JSObject*> glob(cx, JS::CurrentGlobalOrNull(cx));
  JS::Rooted<JSObject*> plain(cx, JS_NewPlainObject(cx));
  JS::Rooted<JSObject*> objectProto(cx, JS::GetRealmObjectPrototype(cx));
  JS::Rooted<JSObject*> global(cx, JS::CurrentGlobalOrNull(cx));
  JS::Rooted<JSObject*> func(
      cx, (JSObject*)JS_NewFunction(cx, (JSNative)1, 0, 0, "dys"));
  JS::Rooted<JSObject*> anon(
      cx, (JSObject*)JS_NewFunction(cx, (JSNative)1, 0, 0, nullptr));
  JS::Rooted<JSFunction*> funcPtr(
      cx, JS_NewFunction(cx, (JSNative)1, 0, 0, "formFollows"));

  JSObject& plainRef = *plain;
  JSFunction& funcRef = *funcPtr;
  JSObject* plainRaw = plain;
  JSObject* funcRaw = func;

  // JS_NewObject will now assert if you feed it a bad class name, so mangle
  // the name after construction.
  char namebuf[20] = "goodname";
  static JSClass cls{namebuf};
  JS::RootedObject badClassName(cx, JS_NewObject(cx, &cls));
  strcpy(namebuf, "\xc7X");

  breakpoint();

  use(glob);
  use(plain);
  use(objectProto);
  use(func);
  use(anon);
  use(funcPtr);
  use(&plainRef);
  use(&funcRef);
  use(JS::GetClass((JSObject*)&funcRef));
  use(plainRaw);
  use(funcRaw);
}

FRAGMENT(JSObject, null) {
  JS::Rooted<JSObject*> null(cx, nullptr);
  JSObject* nullRaw = null;

  breakpoint();

  use(null);
  use(nullRaw);
}

File: js/src/gdb/tests/test-JSObject.py
# Printing JSObjects.
# flake8: noqa: F821

assert_subprinter_registered("SpiderMonkey", "ptr-to-JSObject")
assert_subprinter_registered("SpiderMonkey", "ref-to-JSObject")

run_fragment("JSObject.simple")

# These patterns look a little strange because of prologue.py's 'set print
# address off', which avoids putting varying addresses in the output. After
# the '(JSObject *) ', there is a 'void *' value printing as the empty
# string.

assert_pretty("glob", "(JSObject *)  [object global]")
assert_pretty("plain", "(JSObject *)  [object Object]")
assert_pretty("objectProto", "(JSObject *)  [object Object] used_as_prototype")
assert_pretty("func", '(JSObject *)  [object Function "dys"]')
assert_pretty("anon", "(JSObject *)  [object Function <unnamed>]")
assert_pretty("funcPtr", '(JSFunction *)  [object Function "formFollows"]')

assert_pretty("badClassName", "(JSObject *)  [object \\307X]")

assert_pretty("plainRef", "(JSObject &) @ [object Object]")
assert_pretty("funcRef", '(JSFunction &) @ [object Function "formFollows"]')
